services/command-router/src/main/java/org/eclipse/hono/commandrouter/impl/CommandTargetMapperImpl.java

```diff
@@ -0,0 +1,169 @@
+ * information regarding copyright ownership.
+ *
+ * This program and the accompanying materials are made available under the
+ * terms of the Eclipse Public License 2.0 which is available at
+ * http://www.eclipse.org/legal/epl-2.0
+ *
+ * SPDX-License-Identifier: EPL-2.0
+ *******************************************************************************/
+
+package org.eclipse.hono.commandrouter.impl;
+
+import java.net.HttpURLConnection;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Objects;
+import java.util.concurrent.atomic.AtomicBoolean;
+
+import org.eclipse.hono.adapter.client.registry.DeviceRegistrationClient;
+import org.eclipse.hono.client.ServerErrorException;
+import org.eclipse.hono.client.ServiceInvocationException;
+import org.eclipse.hono.commandrouter.CommandTargetMapper;
+import org.eclipse.hono.deviceconnection.infinispan.client.DeviceConnectionInfo;
+import org.eclipse.hono.tracing.TracingHelper;
+import org.eclipse.hono.util.DeviceConnectionConstants;
+import org.eclipse.hono.util.MessageHelper;
+import org.eclipse.hono.util.RegistrationAssertion;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import io.opentracing.Span;
+import io.opentracing.SpanContext;
+import io.opentracing.Tracer;
+import io.opentracing.tag.Tags;
+import io.vertx.core.Future;
+import io.vertx.core.json.JsonArray;
+import io.vertx.core.json.JsonObject;
+
+/**
+ * A component for mapping an incoming command to the gateway (if applicable)
+ * and protocol adapter instance that can handle it.
+ */
+public class CommandTargetMapperImpl implements CommandTargetMapper {
+
+    private static final Logger LOG = LoggerFactory.getLogger(CommandTargetMapperImpl.class);
+
+    private final Tracer tracer;
+    private final AtomicBoolean initialized = new AtomicBoolean(false);
+    private DeviceRegistrationClient registrationClient;
+    private DeviceConnectionInfo deviceConnectionInfo;
+
+    /**
+     * Creates a new GatewayMapperImpl instance.
+     *
+     * @param tracer The tracer instance.
+     * @throws NullPointerException if tracer is {@code null}.
+     */
+    public CommandTargetMapperImpl(final Tracer tracer) {
+        this.tracer = Objects.requireNonNull(tracer);
+    }
+
+    @Override
+    public void initialize(final DeviceRegistrationClient registrationClient, final DeviceConnectionInfo deviceConnectionInfo) {
+        this.registrationClient = registrationClient;
+        this.deviceConnectionInfo = deviceConnectionInfo;
+        initialized.set(true);
+    }
+
+    @Override
+    public final Future<JsonObject> getTargetGatewayAndAdapterInstance(final String tenantId, final String deviceId,
+            final SpanContext context) {
+        Objects.requireNonNull(tenantId);
+        Objects.requireNonNull(deviceId);
+
+        if (!initialized.get()) {
+            LOG.error("not initialized");
+            return Future.failedFuture(new ServerErrorException(HttpURLConnection.HTTP_INTERNAL_ERROR));
```

Comment:
FMPOV this calls for an `IllegalStateException` (which will then be mapped to an internal error, I believe)